ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/Monitor.hs

```diff
@@ -320,16 +346,28 @@ localRoots actions@PeerSelectionActions{ readLocalRootPeers
                                        }
            policy
            st@PeerSelectionState{
-             bigLedgerPeers,
              localRootPeers,
              publicRootPeers,
              knownPeers,
              establishedPeers,
              activePeers,
              inProgressDemoteHot,
              inProgressDemoteToCold,
-             targets = PeerSelectionTargets{targetNumberOfKnownPeers}
-           } =
+             targets = PeerSelectionTargets{targetNumberOfKnownPeers},
+             ledgerStateJudgement,
+             bootstrapPeersFlag
+           }
+  -- If the current ledger state is TooOld we can only trust our local
+  -- root peers, this means that if we remove any local root peer we
+  -- might no longer abide by the invariant that we are only connected to
+  -- trusted peers. E.g. Local peers = A, B*, C* (* means trusted peer), if
+  -- the node is in bootrstrap mode and decided to reconfigure the local root
+  -- peers to the following set: A*, B, C*, D*, E. Notice that B is no longer
+  -- trusted, however we will keep a connection to it until the outbound
+  -- governor notices it and disconnects from it. Although this eventual
+  -- consistency is okay if we disallow local root reconfigurations we can
+  -- ensure a stronger property in the node, which is desirable.
+  | not (isInSensitiveState bootstrapPeersFlag ledgerStateJudgement) =
```

Comment:
I have thought about this and if we allow the user to reconfigure the local roots, then there can be violations of the invariant that when `hasOnlyBootstrapPeers` is enabled then the known set only has trustable peers. If the user switches a local root from trustable to untrustable then this will happen and although we should demote it quickly I don't like this extra complexity. If bootstrap mode is a sensitive mode and the user has misconfigured its local roots then he is the main responsible for that mistake and he should restart the node with the right configuration. Otherwise I think it is better for us if we strive to keep this strong invariant and not weaken it

Reply:
With this approach the user has no way to fix the system other than restart the node, which is surprising.  For people who will not notice that the local roots has not changed, even though they changed the topology file & send `SIGHUP`; the node will still be running the untrusted peers.  From security stand point this is not right.

I think it's acceptable that local roots will be demoted in reasonable time frame.  I know it makes testing more challenging, but not impossible.  The added complexity will be mostly on that end, which is fine for me.